content/_ext/yaml_gallery_generator.py

```diff
@@ -2,17 +2,98 @@
 from textwrap import dedent
 import pathlib
 
-def build_from_yaml(filename, display_name):
 
-    with open(f'{filename}.yaml') as fid:
-        items = yaml.safe_load(fid)
+def _tag_in_item(item, tag_str):
+    if tag_str is None:
+        return True
+    all_tags = []
+    for k, e in item['tags'].items():
+        all_tags.extend(e)
+    return tag_str in all_tags
+
+
+def _generate_tag_keys(all_items):
+
+    key_set = set()
+    for item in all_items:
+        for k, e in item['tags'].items():
+            key_set.add(k)
+    
+    key_list = list(key_set)
+    key_list.sort()
+
+    return key_list
+
+
+def _generate_tag_set(all_items, tag_key=None):
+
+    tag_set = set()
+    for item in all_items:
+        for k, e in item['tags'].items():
+            if tag_key and k != tag_key:
+                continue            
+            for t in e:
+                tag_set.add(t)
+
+    return tag_set
+
+
+def _sort_tags(tag_set):
+    
+    tag_list = list(tag_set)
+    tag_list.sort()
+    return tag_list
+
+
+def _generate_tag_menu(all_items, tag_key):
+
+    tag_set = _generate_tag_set(all_items, tag_key)
+    tag_list = _sort_tags(tag_set)
+
+    hrefs = ''
+    for tag in tag_list:
+        hrefs = hrefs + f'<a class="dropdown-item" href="/pages/links/{tag.replace(" ", "-")}.html">{tag.title()}</a> \n' 
+
+    tag_menu_html = f"""
+<div class="dropdown">
+<button class="btn btn-sm btn-primary m-2 dropdown-toggle" data-toggle="collapse" data-target="#{tag_key}" aria-haspopup="true">{tag_key.title()}</button>
+<div id="{tag_key}" class="collapse dropdown-menu">
+{hrefs}
+</div>
+</div>
+"""
+    return tag_menu_html
+
+
+def _generate_menu(all_items, flt=None):
+    
+    key_list = _generate_tag_keys(all_items)
+    menu_html='<div class="d-flex flex-row">' + '\n'
+    for tag_key in key_list:
+        menu_html += _generate_tag_menu(all_items, tag_key) + '\n'
+    if flt:
+        menu_html += '<a type="button" class="btn btn-link" href="/pages/links.html">Return to External Links Gallery</a>' + '\n'
```

Comment:
Suggestion: label the button with "Return to Full Gallery"

I just find the "Return to External Link Gallery" a bit jarring since I never actually left the gallery

(sorry to be pedantic)

Reply:
Easy enough change!

Reply:
Awesome, thanks.